vnext/JSI/Shared/ChakraRuntime.h

```diff
@@ -4,53 +4,69 @@
 #pragma once
 
 #include "ChakraObjectRef.h"
-#include "ChakraRuntimeArgs.h"
 
 #include "jsi/jsi.h"
 
-#ifdef CHAKRACORE
-#include "ChakraCore.h"
-#include "ChakraCoreDebugger.h"
-#else
-#ifndef USE_EDGEMODE_JSRT
-#define USE_EDGEMODE_JSRT
-#endif
-#include "jsrt.h"
-#endif
-
 #include <memory>
-#include <mutex>
-#include <sstream>
-
-#if !defined(CHAKRACORE)
-class DebugProtocolHandler {};
-class DebugService {};
-#endif
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
 
 namespace Microsoft::JSI {
 
+struct ChakraRuntimeArgs;
+
+class ChakraPreparedJavaScript : public facebook::jsi::PreparedJavaScript {
+ public:
+  ChakraPreparedJavaScript(
+      std::string &&sourceUrl,
+      const std::shared_ptr<const facebook::jsi::Buffer> &sourceCode,
+      const std::shared_ptr<const facebook::jsi::Buffer> &bytecode)
+      : m_sourceUrl{std::move(sourceUrl)}, m_sourceCode{sourceCode}, m_bytecode{bytecode} {}
+
+  inline const std::string &SourceUrl() const {
+    return m_sourceUrl;
+  }
+
+  inline const std::shared_ptr<const facebook::jsi::Buffer> &SourceCode() const {
+    return m_sourceCode;
+  }
+
+  inline const std::shared_ptr<const facebook::jsi::Buffer> &Bytecode() const {
+    return m_bytecode;
+  }
+
+ private:
+  std::string m_sourceUrl;
+  std::shared_ptr<const facebook::jsi::Buffer> m_sourceCode;
+  std::shared_ptr<const facebook::jsi::Buffer> m_bytecode;
+};
+
+// TODO (yicyao)
 class ChakraRuntime : public facebook::jsi::Runtime {
  public:
-  ChakraRuntime(ChakraRuntimeArgs &&args) noexcept;
+  ChakraRuntime(const std::shared_ptr<ChakraRuntimeArgs> &args) noexcept;
+  void Initialize();
+
   ~ChakraRuntime() noexcept;
 
 #pragma region Functions_inherited_from_Runtime
 
+  // Buffer must contain UTF-8 encoded JavaScript source code.
   facebook::jsi::Value evaluateJavaScript(
       const std::shared_ptr<const facebook::jsi::Buffer> &buffer,
-      const std::string &sourceURL) override;
+      const std::string &sourceUrl) override;
 
+  // The returned std::shared_ptr<const facebook::jsi::PreparedJavaScript>
+  // points to a ChakraPreparedJavaScript and must only be used while this
+  // ChakraRuntime is still alive.
```

Comment:
>alive [](start = 28, length = 5)

I'm a bit confused by this comment. Would you be able to clarify?

Reply:
In ChakraCoreRuntime,, the ChakraPreparedJavaScript returned by prepareJavaScript holds a ChakraCoreByteCodeBuffer, which in turns holds a ChakraObjectRef. (We do so to avoid copying bytecode from JS engined owned memory to RN owned memory.) So the returned ChakraPreparedJavaScript must only be used while the ChakraCoreRuntime that created it is still alive.

---
In reply to: [345454257](https://github.com/microsoft/react-native-windows/pull/3650#discussion_r345454257) [](ancestors = 345454257)